app/src/main/java/org/thoughtcrime/securesms/providers/BlobContentProvider.java

```diff
@@ -0,0 +1,111 @@
+package org.thoughtcrime.securesms.providers;
+
+import android.content.ContentValues;
+import android.database.Cursor;
+import android.net.Uri;
+import android.os.MemoryFile;
+import android.os.ParcelFileDescriptor;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import org.signal.core.util.StreamUtil;
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.dependencies.ApplicationDependencies;
+import org.thoughtcrime.securesms.util.MemoryFileUtil;
+import org.thoughtcrime.securesms.util.Util;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public final class BlobContentProvider extends BaseContentProvider {
+
+  private static final String TAG = Log.tag(BlobContentProvider.class);
+
+  @Override
+  public boolean onCreate() {
+    Log.i(TAG, "onCreate()");
+    return true;
+  }
+
+  @Override
+  public ParcelFileDescriptor openFile(@NonNull Uri uri, @NonNull String mode) throws FileNotFoundException {
+    Log.i(TAG, "openFile() called: " + uri);
+
+    try {
+      try (InputStream stream = BlobProvider.getInstance().getStream(ApplicationDependencies.getApplication(), uri)) {
+        Long fileSize = BlobProvider.getFileSize(uri);
+        if (fileSize == null) {
+          Log.w(TAG, "No file size available");
+          throw new FileNotFoundException();
+        }
+
+        return getParcelStreamForStream(stream, Util.toIntExact(fileSize));
+      }
+    } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)